### PR TITLE
Unit Tests for On Device Sampling

### PR DESCRIPTION
**Pull request by quic-sanising**
- `test_sampler_transform`
    Currently, QEfficient is responsible for export and compiling a model with and without Sampler nodes attached. This unit test confirms the presence of the requisite sampler nodes in the compiled QPC.

### CODE REVIEW

File: tests/transformers/sampler/test_sampler.py
Comment (ochougul) [inline tests/transformers/sampler/test_sampler.py:1]: can we also run the two session with the fixed prompt and make sure the outputs don't match with each other. But match with a golden output that we know matches with pytorch execution as well?

**Review by ochougul — CHANGES_REQUESTED** [review]

**Comment by quic-rishinr** [comment]
@quic-sanising can you add a small feature description under /docs/source/quick_start.md supported feature section? also provide the example script link in the description

**Comment by quic-sanising** [comment]
> @quic-sanising can you add a small feature description under /docs/source/quick_start.md supported feature section? also provide the example script link in the description

Done

**Review by quic-amitraj — CHANGES_REQUESTED** [review]
Please fix lint error.

File: QEfficient/generation/text_generation_inference.py [hunk 1]
```diff
@@ -342,6 +345,11 @@ def cloud_ai_100_exec_kv(
         :Write_io_dir (str): Path to write the input and output files. ``Defaults to None``.
         :automation (bool): If true, it prints input, output, and performance stats. ``Defaults to False``.
         :prompt_to_lora_id_mapping (List[int]): Mapping to associate prompts with their respective LoRA adapter.
+        :include_sampler (bool): Enable/Disable sampling of next tokens.
+        :return_pdfs (bool): Return probability distributions along with sampled
```
[the remark this answers is not in the document]
Comment (quic-sanising) [reply QEfficient/generation/text_generation_inference.py:349]: Done

File: QEfficient/generation/text_generation_inference.py [hunk 2]
```diff
@@ -411,14 +426,59 @@ def __init__(
...
+            if session_input_name in sampler_inputs:
+                count += 1
+                if count == len(sampler_inputs):
+                    break
+        if count == 0:
+            self.include_sampler = False
+        elif count < len(sampler_inputs):
+            raise ValueError(
+                "The provided QPC does not have the required number of inputs to run sampling "
+                f"on the QAIC device (only {count}/{len(sampler_inputs)} inputs provided). Partial "
```
Comment (quic-hemagnih) [inline QEfficient/generation/text_generation_inference.py:465]: I think we should do count % sampler_inputs here. If we divide count by len(sampler_inputs) then it would return 0.

File: QEfficient/generation/text_generation_inference.py [hunk 3]
```diff
@@ -411,14 +426,59 @@ def __init__(
...
+            "temperatures",
+            "top_ks",
+            "top_ps",
+            "min_ps",
+            "random_numbers",
+        ]
+        count = 0
+        for session_input_name in self._session.input_names:
+            if session_input_name in sampler_inputs:
+                count += 1
```
Comment (quic-hemagnih) [inline QEfficient/generation/text_generation_inference.py:457]: Can there be a case where user provides the same session_input_names multiple times. In that case how we will catch it in this code. 
count variable will keep on incrementing and may satisfy the condition

File: QEfficient/generation/text_generation_inference.py [hunk 4]
```diff
@@ -411,14 +426,59 @@ def __init__(
...
+                if count == len(sampler_inputs):
+                    break
+        if count == 0:
```
Comment (quic-hemagnih) [inline QEfficient/generation/text_generation_inference.py:460]: I think we can avoid this if.. else block.
at line 455 by default set self.include_sampler = False.
Then at line 458 before break set it to True.
At line 462 just check for error condition.

File: QEfficient/generation/text_generation_inference.py (hunk 3 above)
Comment (quic-sanising) [reply QEfficient/generation/text_generation_inference.py:457]: `self._session.input_names` come from the exported ONNX file. If there are duplicate names, say `abc`, the ONNX will convert them to something like `abc_0`, `abc_1`, so on. So, we would never get the same name multiple times. 

However, if accuracy is the only priority here and performance is not, I could see `set()` but it would add a slight overhead of `O(n)`.

File: QEfficient/generation/text_generation_inference.py (hunk 2 above)
Comment (quic-sanising) [reply QEfficient/generation/text_generation_inference.py:465]: This is only a print statement. We are not actually dividing here. So, if `count = 5` and `len(sampler_inputs) = 10`, it would print `(only 5/10 inputs provided)`.

File: QEfficient/generation/text_generation_inference.py (hunk 4 above)
Comment (quic-sanising) [reply QEfficient/generation/text_generation_inference.py:460]: In case the user provides `include_sampler` as input, `self.include_sampler` is not set to `False`. That is why, we need the check in line 460. 

We can only avoid the else block in line 468.

**Comment by quic-sanising** [comment]
@quic-amitraj The lint failures were happening because the linter is installing `ruff v0.12.2` whereas the `.pre-commit-config.yaml` file has an older version of `v0.5.2`. 

To fix the errors, we need to either install `ruff v0.5.2` in the linter or update the `.pre-commit-config.yaml` file to version `v0.12.2`.

File: tests/transformers/sampler/test_sampler.py
Comment (quic-sanising) [reply tests/transformers/sampler/test_sampler.py:1]: Please refer to the test `test_random_sampling` below.

File: tests/transformers/sampler/test_sampler.py [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+    pytest.param(
+        "meta-llama/Llama-3.1-8B",  # model
+        Constants.INPUT_STR * 4,  # prompts
+        32,  # prefill_seq_len
+        256,  # ctx_len
+        20,  # generation_len
+        4,  # full_batch_size
+        1,  # spec_length
+    ),
+]
```
Comment (ochougul) [inline tests/transformers/sampler/test_sampler.py:29]: can we use single layer model for this?

**Review by ochougul — CHANGES_REQUESTED** [review]
Everything else LGTM, just see if we can use single layer model in our tests

File: QEfficient/generation/text_generation_inference.py [hunk 6]
```diff
@@ -411,14 +426,59 @@ def __init__(
...
+            self.include_sampler = False
+        elif count < len(sampler_inputs):
+            raise ValueError(
+                "The provided QPC does not have the required number of inputs to run sampling "
+                f"on the QAIC device (only {count}/{len(sampler_inputs)} inputs provided). Partial "
+                "sampling support is not available. Please check the QPC and try again."
+            )
+
+        if include_sampler and not self.include_sampler:
```
Comment (quic-rishinr) [inline QEfficient/generation/text_generation_inference.py:470]: Can we include this check in line 489 itself?

File: QEfficient/generation/text_generation_inference.py [hunk 7]
```diff
@@ -574,6 +641,21 @@ def prepare_decode_inputs(self):
             decode_inputs["position_ids"] = self.decode_pos_ids
         if self.batch_index is not None:
             decode_inputs["batch_index"] = self.batch_index
+        if self.include_sampler:
+            decode_inputs["last_accepted_output_tokens"] = decode_inputs["input_ids"]
+            for op in [
```
Comment (quic-rishinr) [inline QEfficient/generation/text_generation_inference.py:646]: if we can maintain sampler_inputs in constants as mentioned above we should be able to reuse here as well right?

File: QEfficient/generation/text_generation_inference.py [hunk 8]
```diff
@@ -775,10 +894,7 @@ def run_continuous_batching_decode(self, prompt_queue, generation_len):
             outputs = self._session.run(decode_inputs)
 
             # Prepare inputs for next iteration
-            logits = outputs["logits"]
-            if len(logits.shape) == 2:
-                logits = np.expand_dims(logits, 1)
-            next_token_id = logits.argmax(2)
+            next_token_id = self._fetch_next_token_id(outputs)
```
Comment (quic-rishinr) [inline QEfficient/generation/text_generation_inference.py:897]: Since this is inside the decode loop would it create any performance drop? Have you done any tests for checking any performance deviation?

File: tests/transformers/sampler/test_sampler.py [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+        mxint8_kv_cache=True,
+        mxfp6_matmul=True,
+    )
+
+    # Generate texts from prompts
+    tokenizer = load_hf_tokenizer(pretrained_model_name_or_path=model)
+    model_w_sampler_exec_info = model_w_sampler.generate(
+        tokenizer=tokenizer,
+        prompts=prompts,
+        device_id=get_available_device_id(),
```
Comment (quic-rishinr) [inline tests/transformers/sampler/test_sampler.py:167]: you need not pass the device ID. you can keep it as None so auto device picker will pick up the available device ID during execution. Please remove device_id=get_available_device_id() from all the methods below.

File: tests/transformers/sampler/test_sampler.py [hunk 10]
```diff
@@ -0,0 +1,344 @@
...
+# SPDX-License-Identifier: BSD-3-Clause
+#
+# -----------------------------------------------------------------------------
+
+from typing import List
+
+import numpy as np
+import pytest
+
+from QEfficient import QEFFAutoModelForCausalLM as AutoModelForCausalLM
```
Comment (quic-rishinr) [inline tests/transformers/sampler/test_sampler.py:13]: Do we need to import QEFFAutoModelForCausalLM as AutoModelForCausalLM? it would create some unnecessary confusions

File: tests/transformers/sampler/test_sampler.py [hunk 11]
```diff
@@ -0,0 +1,344 @@
...
+        prompts=prompts,
+        device_id=get_available_device_id(),
+        generation_len=generation_len,
+        include_sampler=False,
+        return_pdfs=False,
+        sampling_params=None,
+    )
+
+    # Compare generated texts
+    golden_texts = {
```
Comment (quic-rishinr) [inline tests/transformers/sampler/test_sampler.py:278]: Can we make this test more robust? The goal is to verify whether the random sampler is working as expected right? One idea is to check if the sampler produces the same output when given the same random seed or random values, and different outputs when  the seed or values change. Secondly if we want to compare outputs with and without the sampler cam we use cosine similarity with a threshold instead of exact matching? im open for suggestions

File: tests/transformers/sampler/test_sampler.py (hunk 5 above)
Comment (quic-sanising) [reply tests/transformers/sampler/test_sampler.py]: Changed model to `TinyLlama/TinyLlama-1.1B-Chat-v1.0`

File: QEfficient/generation/text_generation_inference.py [hunk 12]
```diff
@@ -342,6 +345,15 @@ def cloud_ai_100_exec_kv(
         :Write_io_dir (str): Path to write the input and output files. ``Defaults to None``.
         :automation (bool): If true, it prints input, output, and performance stats. ``Defaults to False``.
         :prompt_to_lora_id_mapping (List[int]): Mapping to associate prompts with their respective LoRA adapter.
+        :include_sampler (bool): Enable/Disable sampling of next tokens.
```
[the remark this answers is not in the document]
Comment (quic-sanising) [reply QEfficient/generation/text_generation_inference.py]: Done

File: QEfficient/generation/text_generation_inference.py [hunk 13]
```diff
@@ -411,14 +426,59 @@ def __init__(
...
         self.is_tlm = is_tlm
+        self.include_sampler = include_sampler
+        self.return_pdfs = return_pdfs
+        self.sampling_params = sampling_params
 
         # Load QPC
         self._session = QAICInferenceSession(qpc_path, device_id, enable_debug_logs=enable_debug_logs)
 
+        # Validate sampler inputs for On-Device Sampling
+        sampler_inputs = [
```
[the remark this answers is not in the document]
Comment (quic-sanising) [reply QEfficient/generation/text_generation_inference.py]: Done

File: QEfficient/generation/text_generation_inference.py (hunk 7 above)
Comment (quic-sanising) [reply QEfficient/generation/text_generation_inference.py]: Yes, done

File: tests/transformers/sampler/test_sampler.py (hunk 9 above)
Comment (quic-sanising) [reply tests/transformers/sampler/test_sampler.py]: Done

File: QEfficient/generation/text_generation_inference.py (hunk 6 above)
Comment (quic-sanising) [reply QEfficient/generation/text_generation_inference.py]: Could you clarify this?

File: tests/transformers/sampler/test_sampler.py (hunk 11 above)
Comment (quic-sanising) [reply tests/transformers/sampler/test_sampler.py:268]: Thanks for the suggestion! The current test checks for exact output matches, which ensures reproducibility with a fixed seed. As for using cosine similarity or other metrics like perplexity or rogue score, they are useful for semantic comparison but can be vague as there are no clear-defined or universally accepted thresholds. For this test, exact matching keeps things deterministic and easier to maintain. Happy to explore alternatives...

File: QEfficient/generation/text_generation_inference.py [hunk 14]
```diff
@@ -411,14 +428,25 @@ def __init__(
...
         write_io_dir: Optional[str] = None,
         is_tlm: Optional[int] = None,
+        include_sampler: bool = False,
+        return_pdfs: bool = False,
+        sampling_params: Optional[Dict[str, Any]] = None,
     ) -> None:
         self._ctx_len = ctx_len
         self._write_io_dir = write_io_dir
         self.is_tlm = is_tlm
+        self.include_sampler = include_sampler
```
Comment (quic-rishinr) [inline QEfficient/generation/text_generation_inference.py:438]: do we need this? as we are assigning self.include_sampler =  validate_sampler_inputs() on line number 446

File: QEfficient/generation/text_generation_inference.py [hunk 15]
```diff
@@ -411,14 +428,25 @@ def __init__(
...
+        self.return_pdfs = return_pdfs
+        self.sampling_params = sampling_params
 
         # Load QPC
         self._session = QAICInferenceSession(qpc_path, device_id, enable_debug_logs=enable_debug_logs)
 
+        # Validate sampler inputs for On-Device Sampling
+        self.include_sampler = validate_sampler_inputs(
```
Comment (quic-rishinr) [inline QEfficient/generation/text_generation_inference.py:446]: can we have something like self.include_sampler = validate_sampler_inputs(set(self._session.input_names), include_sampler if include_sampler is not False else False

File: tests/transformers/sampler/test_sampler.py [hunk 16]
```diff
@@ -0,0 +1,334 @@
...
+        mxint8_kv_cache=True,
+        mxfp6_matmul=True,
+    )
+
+    # Generate texts from prompts
+    tokenizer = load_hf_tokenizer(pretrained_model_name_or_path=model)
+    model_w_sampler_exec_info = model_w_sampler.generate(
+        tokenizer=tokenizer,
+        prompts=prompts,
+        device_id=None,
```
Comment (quic-rishinr) [inline tests/transformers/sampler/test_sampler.py:157]: you can remove the device ID completely from test. Anyway default value is None

**Review by quic-rishinr — CHANGES_REQUESTED** [review]

File: QEfficient/generation/text_generation_inference.py (hunk 14 above)
Comment (quic-sanising) [reply QEfficient/generation/text_generation_inference.py]: Right, removed.

File: QEfficient/generation/text_generation_inference.py (hunk 15 above)
Comment (quic-sanising) [reply QEfficient/generation/text_generation_inference.py:445]: I don't think we need to do this. `include_sampler` is a mandatory boolean variable with a default value of `False`. The only other possible value is `True`. Both of these scenarios are handled in `validate_sampler_inputs()` function.

File: QEfficient/generation/text_generation_inference.py (hunk 8 above)
Comment (quic-sanising) [reply QEfficient/generation/text_generation_inference.py:848]: I still do not follow. Are you talking about this reshape?
```py
return outputs["next_tokens"].reshape(outputs["next_tokens"].shape[0], outputs["next_tokens"].shape[1])
```

If yes, this converts 3D tensor of shape `(batch_size, 1, 1)` to a 2D tensor of shape `(batch_size, 1)`. There is no operation that can cause a drop in performance. 

If you are still curious about the performance gains, please reach out to me for a complete performance report.

File: .pre-commit-config.yaml [hunk 17]
```diff
@@ -1,7 +1,7 @@
 repos:
   - repo: https://github.com/astral-sh/ruff-pre-commit
     # Ruff version.
-    rev: v0.5.2
+    rev: v0.12.7
```
Comment (ochougul) [inline .pre-commit-config.yaml:4]: why?

File: examples/on_device_sampling.py
Comment (ochougul) [inline examples/on_device_sampling.py:1]: are the print statements intentional here?

Comment (quic-sanising) [reply examples/on_device_sampling.py:1]: What print statements?

**Comment by quic-hemagnih** [comment]
Anything pending for this PR? @ochougul  @quic-sanising

**Comment by quic-sanising** [comment]
> Anything pending for this PR? @ochougul @quic-sanising

Nothing from my end. Please go ahead with the merge if the CI has passed.

File: QEfficient/cloud/infer.py [hunk 18]
```diff
@@ -231,7 +231,6 @@ def main(
             tokenizer,
             prompts=prompt,
             device_id=device_group,
-            prompt=prompt,
```
[the remark this answers is not in the document]
Comment (quic-rishinr) [reply QEfficient/cloud/infer.py:238]: As rightly shown update the infer to follow the same.

**Review by ochougul — APPROVED** [review]

**Review by quic-hemagnih — APPROVED** [review]
